File: rust/src/expires.rs
//! Expiration time helpers.
//!
//! Convenience functions for generating ISO 8601 expiration timestamps.

use time::format_description::well_known::Rfc3339;
use time::{Duration, OffsetDateTime};

pub fn seconds(n: u64) -> String {
    offset(Duration::seconds(n as i64))
}

pub fn minutes(n: u64) -> String {
    offset(Duration::minutes(n as i64))
}

pub fn hours(n: u64) -> String {
    offset(Duration::hours(n as i64))
}

pub fn days(n: u64) -> String {
    offset(Duration::days(n as i64))
}

pub fn weeks(n: u64) -> String {
    offset(Duration::weeks(n as i64))
}

fn offset(duration: Duration) -> String {
    let dt = OffsetDateTime::now_utc() + duration;
    dt.format(&Rfc3339)
        .unwrap_or_else(|_| "1970-01-01T00:00:00Z".to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn minutes_format() {
        let result = minutes(5);
        assert!(result.contains('T'));
    }

    #[test]
    fn hours_later_than_minutes() {
        let m = minutes(1);
        let h = hours(1);
        assert!(h > m);
    }
}

File: go/expires.go
package mpp

import "time"

// Seconds returns an RFC3339 timestamp offset from now by n seconds.
func Seconds(n uint64) string { return offset(time.Duration(n) * time.Second) }

// Minutes returns an RFC3339 timestamp offset from now by n minutes.
func Minutes(n uint64) string { return offset(time.Duration(n) * time.Minute) }

// Hours returns an RFC3339 timestamp offset from now by n hours.
func Hours(n uint64) string { return offset(time.Duration(n) * time.Hour) }

// Days returns an RFC3339 timestamp offset from now by n days.
func Days(n uint64) string { return offset(time.Duration(n) * 24 * time.Hour) }

// Weeks returns an RFC3339 timestamp offset from now by n weeks.
func Weeks(n uint64) string { return offset(time.Duration(n) * 7 * 24 * time.Hour) }

func offset(duration time.Duration) string {
	return time.Now().UTC().Add(duration).Format(time.RFC3339)
}
